src/LeaseManager.cc

```diff
@@ -0,0 +1,307 @@
+
+            // Keep track of the largest allocated leaseId and only auto "renew"
+            // the other leaseIds.
+            if (maxAllocatedLeaseId == 0) {
+                maxAllocatedLeaseId = leaseId;
+                continue;
+            } else if (leaseId > maxAllocatedLeaseId) {
+                uint64_t temp = maxAllocatedLeaseId;
+                maxAllocatedLeaseId = leaseId;
+                leaseId = temp;
+            }
+
+            uint64_t leaseTerm = clock.getTime() + LeaseCommon::LEASE_TERM_US;
+            leaseMap[leaseId] = leaseTerm;
+            expirationOrder.insert({leaseTerm, leaseId});
+        } catch (std::invalid_argument& e) {
+            LOG(ERROR, "Unable to recover lease: %s", object.name);
+        }
+    }
+
+    if (maxAllocatedLeaseId > 0) {
+        lastIssuedLeaseId = maxAllocatedLeaseId - 1;
+    }
+}
+
+/**
+ * Attempts to renew a lease with leaseId. If the requested leaseId is still
+ * live the lease is renewed and its term is extended; if not a new leaseId is
+ * issued and returned.
+ *
+ * \param leaseId
+ *      The leaseId that a client wishes to renew if possible.  Note that
+ *      leaseId = 0 is implicitly invalid and will always cause a new lease to
+ *      be returned.
+ * \return
+ *      leaseId and leaseTerm for the renewed lease or new lease.
+ */
+WireFormat::ClientLease
+LeaseManager::renewLease(uint64_t leaseId)
+{
+    Lock lock(mutex);
+    WireFormat::ClientLease clientLease = renewLeaseInternal(leaseId, lock);
+
+    // Poke the preallocator if we are getting close to running out of leases.
+    if (maxAllocatedLeaseId - lastIssuedLeaseId < PREALLOCATION_LIMIT / 4) {
+        preallocator.start(0);
+    }
+
+    return clientLease;
+}
+
+/**
+ * Start background timers to perform preallocation and cleaning and also start
+ * the cluster clock updater.
+ */
+void
+LeaseManager::startUpdaters()
+{
+    preallocator.start(0);
+    cleaner.start(0);
+    clock.startUpdater();
+}
+
+/**
+ * Constructor for the LeasePreallocator.
+ * \param context
+ *      Overall information about the RAMCloud server and provides access to
+ *      the dispatcher.
+ * \param leaseManager
+ *      Provides access to the leaseManager to perform preallocation.
+ */
+LeaseManager::LeasePreallocator::LeasePreallocator(Context* context,
+                                                  LeaseManager* leaseManager)
+    : WorkerTimer(context->dispatch)
+    , leaseManager(leaseManager)
+{}
+
+/**
+ * The handler performs the lease preallocation.
+ */
+void
+LeaseManager::LeasePreallocator::handleTimerEvent()
+{
+    uint64_t preallocationCount = 0;
+    while (true) {
+        LeaseManager::Lock lock(leaseManager->mutex);
+        preallocationCount = leaseManager->maxAllocatedLeaseId -
+                             leaseManager->lastIssuedLeaseId;
+        if (preallocationCount >= PREALLOCATION_LIMIT)
+            break;
+        leaseManager->allocateNextLease(lock);
```

Comment:
Let's discuss recommended alternatives (at least for before we have our next project up and running ;-) )